File: C++/CreationalDesignPattern/SingletonDesignPattern.cpp
#include <iostream>
#include <memory>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

class Logger {
    static std::unique_ptr<Logger> instance;
    static std::mutex mtx;
    std::vector<std::string> logs;

    Logger() { std::cout << "object created" << std::endl; }

public:
    Logger(const Logger&) = delete;
    Logger& operator=(const Logger&) = delete;

    void appendLogs(const std::string& log) { logs.push_back(log); }

    static Logger& getInstanceNormalThreadUnsafe() {
        if (!instance) {
            instance = std::unique_ptr<Logger>(new Logger());
        }
        return *instance;
    }

    static Logger& getInstanceSynchronized() {
        std::lock_guard<std::mutex> lock(mtx);
        if (!instance) {
            instance = std::unique_ptr<Logger>(new Logger());
        }
        return *instance;
    }

    static Logger& getInstance() {
        if (!instance) {
            std::lock_guard<std::mutex> lock(mtx);
            if (!instance) {
                instance = std::unique_ptr<Logger>(new Logger());
            }
        }
        return *instance;
    }
};

std::unique_ptr<Logger> Logger::instance = nullptr;
std::mutex Logger::mtx;

int main() {
    // Logger& logger1 = Logger::getInstance();
    // Logger& logger2 = Logger::getInstance();
    // Logger& logger3 = Logger::getInstance();

    auto task1 = []() { Logger::getInstance(); };

    // std::thread thread1(task1);
    // std::thread thread2(task1);
    // thread1.join();
    // thread2.join();

    std::vector<std::thread> threads;
    for (int i = 0; i < 2; i++) {
        threads.emplace_back(task1);
    }
    for (auto& t : threads) {
        t.join();
    }

    return 0;
}

File: Python/CreationalDesignPattern/SingletonDesignPattern.py
import threading
from typing import List, Optional


class Logger:
    _instance: Optional["Logger"] = None
    _lock = threading.RLock()

    def __init__(self):
        print("object created")
        self.logs: List[str] = []

    def append_logs(self, log: str):
        self.logs.append(log)

    @classmethod
    def get_instance_normal_thread_unsafe(cls) -> "Logger":
        if cls._instance is None:
            cls._instance = Logger()
        return cls._instance

    @classmethod
    def get_instance_synchronized(cls) -> "Logger":
        with cls._lock:
            if cls._instance is None:
                cls._instance = Logger()
            return cls._instance

    @classmethod
    def get_instance(cls) -> "Logger":
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = Logger()
        return cls._instance


def main():
    # logger1 = Logger.get_instance()
    # logger2 = Logger.get_instance()
    # logger3 = Logger.get_instance()

    def task1():
        Logger.get_instance()

    # thread1 = threading.Thread(target=task1)
    # thread2 = threading.Thread(target=task1)
    # thread1.start()
    # thread2.start()

    for _ in range(2):
        threading.Thread(target=task1).start()


if __name__ == "__main__":
    main()
